### DISCUSSION TITLE
Generating a module block that includes for_each iteration

### DISCUSSION THREAD

**Discussion by ottramst** [terramate-io/terramate#660] [Q&A] [answered]
Hi again, everyone!

Need some advice or best practise tips on how it would be best to configure code generation for a module that is required to be iterated.

For example:
```
globals {
  items = [
    {
      name = "foobar"
    }
  ]
}

generate_hcl "_module.tf" {
  content {
    module "this" {
      for_each = {
        for item in var.items : item.name => item
      }

      source = "..."

      name = each.value.name
    }
  }
}
```

One solution would be to pass the `global.items` to a Terraform variable and use that, but maybe there is a cleaner way for this?

Thanks in advance!

**Comment by katcipis** [comment, marked as answer]
You could do the for each evaluation itself using a `lets` block and then just reference the results on the for_each. Today it is a limitation that you can't have a obj/list comprehension when partial evaluation is involved (which happens inside the content block), but the `lets` block is fully evaluated (like `globals`). So this:

```hcl
globals {
  items = [
    {
      name = "foobar"
    }
  ]
}

generate_hcl "_module.tf" {
  lets {
      item_map = {
        for item in global.items : item.name => item
      }
  }
  content {
    module "this" {
      for_each = let.item_map

      source = "some.source"
      name = each.value.name
    }
  }
}
```

Will generate this:

```hcl
module "this" {
  for_each = {
    foobar = {
      name = "foobar"
    }
  }
  name   = each.value.name
  source = "some.source"
}
```

Which should work. Let me know if I got something wrong or this was helpful.

**Reply by ottramst** [reply]
Thank you! This is much cleaner.
